src/lib/coda_intf/core_intf.ml

```diff
@@ -35,6 +35,8 @@ module type Ledger_proof_generalized_intf = sig
     sig
       module V1 : sig
         type t [@@deriving sexp, bin_io, yojson, version]
+
+        val statement : t -> transaction_snark_statement
```

Comment:
So I am a little confused with this. @psteckler suggested that the functor arguments be specific versions and not versioned modules so that the choice of version happens at the functor application. Now If I were to pass `Ledger_proof.Stable.V1` to `Transaction_snark_work.Make` then I'd need the version specific functions, in this case `statement`? and when the version changes and we do `Transaction_snark_work.Make (Ledger_proof.Stable.V2)` we'd need `V2` specific functions?

Reply:
Now that I think about it, versioned types seem like they should always live outside functors and take type parameters for types they would be functored over. Consider:

```ocaml
module A = struct
  module Stable = struct
    module V1 = ...
    module V2 = ...
  end
end

module M (B : ...) = struct
  module Z = struct
    module Stable = struct
      module V1 = ...
      module V2 = ...
    end
  end
end
```
If `M` takes a version of `A` as an argument, and `M.Z` depends on it, then we invent some new versions and prevent from converting between others
* creating `A.V2` before `M.Z.V2` isolates the valid `M(A.V1).Z.V1` in a different module (and thus versioning hierarchy) from `M(A.V2).Z.V1` and `M(A.V2).Z.V2`, as well as inventing `M(A.V1).Z.V2`
* flipping the order swaps which is isolated and which is invented
* changing both at the same time isolates the new combination from *all previous real versions* and invents 2 new versions
* any other versioned type can use `M(A.Latest).Z.V2` as if it were a versioned type, even though later versions of `A` would change the underlying structure

I don't think there's any good way to make versioning in functors make sense, so avoiding it would be better. (/cc @psteckler, if you agree, we should probably check we're not in a functor in the ppx)

This also means that we shouldn't need to put any code inside versioned types.